keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,482 @@
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place
+- [ ] (R) Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [etcd-io/website], for publication to [etcd.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[etcd.io]: https://etcd.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[etcd-io/website]: https://github.com/etcd-io/website
+
+## Summary
+
+We are introducing functionalities to determine if a feature is enabled for the whole etcd cluster. Users would be able to query cluster feature enablement in a consistent way on the cluster level. 
+
+## Motivation
+
+In [KEP-4578](https://github.com/kubernetes/enhancements/issues/4578) we introduced the feature gate framework into etcd. But the feature gate only works on the local server level. Different members could hold different values of a feature enablement. A request may go through completely different code paths depending on if a feature is enabled/disabled on a server. This is super critical for data consistency if the request is touching etcd data in anyway. Such features requires all members to agree on the feature enablement together, so we need to add functionalities to the feature gate to work on the cluster level for these cluster level features. 
+
+### Goals
+
+* Establish a mechanism to safely and consistently determine if a cluster-level feature is enabled/disabled on all members.
+
+* Introduce apis/methods to query the enablement state of a cluster level feature.
+
+### Non-Goals
+
+* Cannot backport a feature gate to a patch version.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+A developer is adding a new feature of a new write type. Because it requires a consensus write, the request should be rejected if the feature is not enabled or the request can be processed on all cluster members. 
+Today, any new raft request type would need to be added one minor version before the actual feature can be enabled, and the leader would have to optimistically assume the feature is enabled or disabled consistently across the cluster.
```

Comment:
```suggestion
Today, each change in etcd API, requires development spenning at least 2 minor versions. First a raft proto change needs to be added and wait a release before it can be used. Only a release later it can be actually used to prevent a incompatible proto being used during upgrade/downgrade process. 
```